### PR TITLE
A0-4577: Implement era abft performance calculation

### PR DESCRIPTION
**Pull request by mike1729**
# Description

Calculate a performance stat for every finalizer at the end of a session and emit an event that a validator is underperforming when a threshold is met. In future, implement banning such validator.

## Type of change

- New feature (non-breaking change which adds functionality)

### CODE REVIEW

**Review by Marcin-Radecki — APPROVED** [review]
Changes looks good. We need migration for `committeeManagement.currentAndNextSessionValidatorStorage` which was discussed offline and task was created for it.

File: pallets/committee-management/src/manager.rs [hunk 1]
```diff
@@ -142,6 +143,8 @@ where
             "Result of clearing the `SessionValidatorBlockCount`, {:?}",
             result.deconstruct()
         );
+
+        C::AbftScoresProvider::clear_scores_for_session(end_index);
```
[the remark this answers is not in the document]
Comment (mike1729) [reply pallets/committee-management/src/manager.rs:147]: Storage is precious so we need to clear it periodically. Remember that storage is kept per block so you can read past values until they're pruned.

[the remark this answers is not in the document]
Comment (mike1729) [reply pallets/committee-management/src/manager.rs:147]: We could clear them on era start, do you think that would be ok?

File: pallets/committee-management/src/impls.rs [hunk 2]
```diff
@@ -345,18 +347,49 @@ impl<T: Config> Pallet<T> {
...
+
+        let is_underperforming = |score| score > minimal_expected_performance;
+
+        let finalizers_perf = T::AbftScoresProvider::scores_for_session(session_id)
+            .map(|score| score.points)
+            .unwrap_or(vec![minimal_expected_performance; finalizers.len()])
+            .into_iter()
+            .map(is_underperforming);
+
+        // TODO make sure ordering is correct on validators list
```
Comment (timorleph) [inline pallets/committee-management/src/impls.rs:369]: This TODO probably shouldn't be here.

If it's about the ABFT order agreeing with the one from the pallets then it should, it should just be the source of that order, unless there is something unnecessarily shuffling it around somewhere in between.

File: pallets/committee-management/src/lib.rs [hunk 3]
```diff
@@ -249,19 +268,65 @@ pub mod pallet {
...
+            underperformed_session_count_threshold: Option<u32>,
+            ban_period: Option<EraIndex>,
+        ) -> DispatchResult {
+            ensure_root(origin)?;
+
+            let mut current_committee_ban_config = FinalityBanConfig::<T>::get();
+
+            if let Some(minimal_expected_performance) = minimal_expected_performance {
+                ensure!(
+                    minimal_expected_performance <= 20,
```
Comment (timorleph) [inline pallets/committee-management/src/lib.rs:287]: What is this 20 about?

**Review by timorleph — APPROVED** [review]
Couple minor comments.

File: pallets/committee-management/src/impls.rs [hunk 4]
```diff
@@ -104,30 +106,26 @@ fn select_committee_inner<AccountId: Clone + PartialEq>(
...
-    let mut finality_committee = choose_finality_committee(
+    let mut finalizers = choose_finality_committee(
         &reserved_committee,
         &non_reserved_committee,
         non_reserved_finality_seats,
         current_session as usize,
     );
 
-    let mut block_producers = match (reserved_committee, non_reserved_committee) {
+    let mut producers = match (reserved_committee, non_reserved_committee) {
```
[the remark this answers is not in the document]
Comment (mike1729) [reply pallets/committee-management/src/impls.rs:116]: I wanted the naming to be consistent.

File: pallets/committee-management/src/lib.rs (hunk 3 above)
Comment (mike1729) [reply pallets/committee-management/src/lib.rs:287]: I wanted to have some check on the value and I don't expect that lagging more than 20 rounds will be considered acceptable.

Comment (mike1729) [reply pallets/committee-management/src/lib.rs:287]: Changed to value representing not producing units all session

File: pallets/committee-management/src/manager.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (mike1729) [reply pallets/committee-management/src/manager.rs]: Done.

File: pallets/committee-management/src/impls.rs (hunk 2 above)
Comment (mike1729) [reply pallets/committee-management/src/impls.rs]: I've removed the comment.

**Pull request merged**